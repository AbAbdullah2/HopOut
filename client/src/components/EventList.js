import React, { Component } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { solid } from '@fortawesome/fontawesome-svg-core/import.macro'

class EventList extends Component {
  state = {
    events: [
      {
        event: {
          _id: 1,
          title: 'HopOut Launch',
          start: '2023-02-22T20:00:00Z',
          end: '2023-02-22T22:00:00Z',
          description: 'Launching HopOut',
          image: 'https://picsum.photos/200/200',
          location: '123 Main St, New York, NY 10001',
          visibility: 'public',
        },
        organizer: {
          name: "HopOut Dev",
        }
      },
      {
        event: {
          _id: 2,
          title: 'HopOut Sale',
          start: '2023-03-22T20:00:00Z',
          end: '2023-03-22T22:00:00Z',
          description: 'Selling HopOut :(',
          image: 'https://picsum.photos/200/200',
          location: '123 Main St, New York, NY 10001',
          visibility: 'public',
        },
        organizer: {
          name: "HopOut CEO",
        }
      },
    ],
  };

  render() {
    let { events } = this.state;

    return (
        <div className='mt-5 w-4/5'>
          <p className='text-4xl text-center'>Event List</p>
          {
            events.map((event) => {
              const start = new Date(event.event.start);
              const end = new Date(event.event.end);
              return (
                <div className='my-2 bg-slate-400 flex' key={event.event._id}>
                  <img className='m-2 w-3/12' src={event.event.image} alt={event.event.title} />
                  <div className='w-9/12 items-center justify-center text-center m-2 space-y-2'>
                    <p className='text-4xl'>{event.event.title}</p>
                    <p><FontAwesomeIcon icon={solid('calendar')} /> {start.toDateString() + ' ' + start.getHours() + ':' + start.getMinutes() + ' - ' + end.getHours() + ':' + end.getMinutes()}</p>
                    <p><FontAwesomeIcon icon={solid('location-dot')} /> {event.event.location}</p>
                    <p>{event.event.description}</p>
                    <p>Organized by {event.organizer.name}</p>
                    <p>Visibility: {event.event.visibility}</p>
                  </div>
                </div>
              );
            })
          }
        </div>
    );
  }
}

export default EventList;