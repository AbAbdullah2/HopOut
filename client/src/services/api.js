import axios from 'axios';

const BASE_URL =
  !process.env.NODE_ENV || process.env.NODE_ENV === 'development'
    ? 'http://localhost:6002'
    : 'https://hopoutapi.herokuapp.com';

// Get all users
async function getAllUsers() {
  const response = await axios.get(`${BASE_URL}/users`).catch(function (error) {
    console.log(error);
  });
  return response;
}

// Get user by id
async function getUser(userId) {
  const response = await axios
    .get(`${BASE_URL}/users/${userId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// create user / sign up / register
// user should be object with fields email, name password
async function register(user) {
  const response = await axios
    .post(`${BASE_URL}/register`, user)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// Delete user / account 
async function deleteAccount(id) {
  const response = await axios
    .delete(`${BASE_URL}/users/${id}`, id)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}


async function emailVerification(user) {
  const response = await axios.post(`${BASE_URL}/verification`, user)
    .catch(function (error) {
      console.log(error);
  });
  return response;
}

async function updateUser(user) {
  const response = await axios
    .put(`${BASE_URL}/users/${user._id}`, user)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// login creds should be object with fields email and password
async function postLogin(creds) {
  const response = await axios
    .post(`${BASE_URL}/login`, creds)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// delete a user (must have field _id)
async function deleteUser(user) {
  const response = await axios
    .delete(`${BASE_URL}/users/${user._id}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function createNewEvent(event) {
  const response = await axios
    .post(`${BASE_URL}/events`, event)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// get all public events
async function getAllPublicEvents() {
  const response = await axios
    .get(`${BASE_URL}/events`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}
// get all private events
async function getAllPrivateEvents(userId) {
  const response = await axios
    .get(`${BASE_URL}/users/privateEvents/${userId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// get all hosted events
async function getAllHostedEvents(userId) {
  const response = await axios.get(`${BASE_URL}/users/hostedEvents/${userId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function getAllAttendedEvents(userId) {
  const response = await axios.get(`${BASE_URL}/users/attendedEvents/${userId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function getEvent(eventId) {
  const response = await axios
    .get(`${BASE_URL}/events/${eventId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function updateEvent(event) {
  const response = await axios
    .put(`${BASE_URL}/events/${event._id}`, event)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function sendFriendReq(senderId, receiverId) {
  const response = await axios
    .put(`${BASE_URL}/friends/sendRequest`, { senderId, receiverId })
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function acceptFriendReq(acceptorId, requesterId) {
  const response = await axios
    .put(`${BASE_URL}/friends/acceptRequest`, { acceptorId, requesterId })
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function declineFriendReq(declinerId, requesterId) {
  const response = await axios
    .put(`${BASE_URL}/friends/declineRequest`, { declinerId, requesterId })
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function removeFriendReq(removerId, otherId) {
  const response = await axios.put(`${BASE_URL}/friends/removeRequest`, { removerId, otherId })
  .catch(function (error) {
    console.log(error);
  });
  return response;
}

async function removeFriend(removerId, friendId) {
  const response = await axios
    .put(`${BASE_URL}/friends/removeFriend`, { removerId, friendId })
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function deleteEvent(eventId) {
  const response = await axios
    .delete(`${BASE_URL}/events/${eventId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function sendInvite(evid, invid) {
  const body = { eventId: evid, inviteeId: invid };
  const response = await axios
    .put(`${BASE_URL}/rsvp/sendInvite`, body)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function unsendInvite(evid, invid) {
  const body = { eventId: evid, uninviteeId: invid };
  const response = await axios.put(`${BASE_URL}/rsvp/unsendInvite`, body)
  .catch(function (error) {
    console.log(error);
  });
  return response;
}

async function rsvpToEvent(userId, eventId) {
  const body = { senderId: userId, eventId };
  const response = await axios
    .put(`${BASE_URL}/rsvp/sendRSVP`, body)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

async function cancelRsvp(userId, eventId) {
  const body = { removerId: userId, eventId };
  const response = await axios
    .put(`${BASE_URL}/rsvp/removeRSVP`, body)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// get all of user's chats by userid
async function getAllChats(userId) {
  const response = await axios
    .get(`${BASE_URL}/getAllChats/${userId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// get all of user's chats by userid
async function getChat(chatId) {
  const response = await axios
    .get(`${BASE_URL}/getChat/${chatId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// send message to chat
async function sendMessage(chatId, senderId, receiverId, message) {
  const body = { chatId, senderId, receiverId, message };
  const response = await axios
    .post(`${BASE_URL}/message/`, body)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// create new chat between two users
async function createChat(person1, person2) {
  const body = { person1, person2 };
  const response = await axios
    .post(`${BASE_URL}/chat/`, body)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// create comment section for event
async function createCommentSection(eventId) {
  const response = await axios
    .post(`${BASE_URL}/commentSection/${eventId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// Post comment
async function createComment(eventId, commentSectionId, sender, message) {
  const body = {
    eventId,
    commentSectionId,
    sender,
    message,
  };
  const response = await axios
    .post(`${BASE_URL}/comment/${commentSectionId}`, body)
    .catch(function (error) {
      console.log(error);
  });
  return response;
}

// get all of events's comments by eventid
async function getAllComments(eventId) {
  const response = await axios
    .get(`${BASE_URL}/getAllComments/${eventId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// delete comment from comment section 
async function deleteComment(commentSectionId, commentId, senderId) {
  const body = { senderId, commentId };

  const response = await axios
    .delete(`${BASE_URL}/deleteComment/${commentSectionId}`, {data: body})
    .catch(function (error) {
      console.log(error);
    });
  return response;
}

// delete entire comment section 
async function deleteCommentSection(eventId) {
  const response = await axios
    .delete(`${BASE_URL}/deleteCommentSection/${eventId}`)
    .catch(function (error) {
      console.log(error);
    });
  return response;
}


export {
  sendInvite,
  unsendInvite,
  rsvpToEvent,
  cancelRsvp,
  getAllUsers,
  getUser,
  register,
  deleteAccount,
  postLogin,
  deleteUser,
  getAllPublicEvents,
  getAllPrivateEvents,
  getEvent,
  createNewEvent,
  updateEvent,
  deleteEvent,
  sendFriendReq,
  acceptFriendReq,
  declineFriendReq,
  removeFriendReq,
  removeFriend,
  updateUser,
  getAllHostedEvents, 
  getAllAttendedEvents,
  emailVerification,
  getAllChats,
  getChat,
  sendMessage,
  createChat,
  createCommentSection,
  createComment,
  getAllComments,
  deleteComment,
  deleteCommentSection,
};

export const host = BASE_URL;
