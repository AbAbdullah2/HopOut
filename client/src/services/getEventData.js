import eventData from '../assets/eventData.js';

export default function getEventData() {
  return eventData;
}