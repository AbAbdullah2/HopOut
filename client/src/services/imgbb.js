import axios from 'axios';
const API_KEY = process.env.REACT_APP_IMGBB_KEY;
const BASE_URL = `https://api.imgbb.com/1/upload?expiration=600&key=${API_KEY}`;
// You should never save API key directly in source code

// Upload to imgbb 
async function uploadImg(img) {
    const form = new FormData();
    form.append("image", img)

    const response = await axios.post(`${BASE_URL}`, form)
    .catch(function (error) {
        console.log(error);
        });
    return response;
}

export { uploadImg  }
  