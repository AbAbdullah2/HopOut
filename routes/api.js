const express = require('express');
const router = express.Router();
const Hello = require('../models/hello');

router.get('/hellos', (req, res, next) => {
  Hello.find({}, 'text')
    .then((data) => res.json(data))
    .catch(next);
});

router.post('/hellos', (req, res, next) => {
  if (req.body.text) {
    Hello.create(req.body)
      .then((data) => res.json(data))
      .catch(next);
  } else {
    res.json({
      error: 'Empty text invalid',
    });
  }
});

router.get('/hellos/:id', (req, res, next) => {
  Hello.findOneAndDelete({ _id: req.params.id })
    .then((data) => res.json(data))
    .catch(next);
});

module.exports = router;