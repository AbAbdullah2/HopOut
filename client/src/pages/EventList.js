import React, { useState, useEffect } from 'react';
import Header from '../components/Header';
import EventCard from '../components/EventCard';
import CategoryFilter from '../components/CategoryFilter';
import { getAllPublicEvents, getAllPrivateEvents } from '../services/api';
import { useNavigate } from 'react-router-dom';
import FriendFilter from '../components/FriendFilter';
import Map from '../components/Map';
import { Switch } from '@headlessui/react'

export function EventList(props) {
  const { curUser, setCurUser} = props;
  const [eventList, setEventList] = useState([]);
  const [privateEventList, setPrivateEventList] = useState([]);

  const [listActive, setListActive] = useState(false);
  const [friendEventsActive, setFriendEventsActive] = useState(false);
  const [selectedFilters, setSelectedFilters] = useState([]);

  const [friendFilters, setFriendFilters] = useState([]);

  const navigate = useNavigate();
  useEffect(() => {
    if (curUser == null) navigate('/login');
    getAllPublicEvents().then((res) => {
      setEventList(res.data.data);
    })

  }, [curUser, navigate]);

  useEffect(() => {
    getAllPrivateEvents(curUser._id).then((resp) => {
      setPrivateEventList(resp.data.data);
    });
  }, [curUser, navigate]);

  const toDisplayEvent = (ev) => {
    let filtered = false;
    if (selectedFilters.length === 0 && friendFilters.length === 0) {
      if (!friendEventsActive) {
        return true;
      }
      if (friendEventsActive && curUser.friends.map((ev) => ev.user).includes(ev.organizer)) {
        return true;
      }
    }

    if (friendEventsActive && !(curUser.friends.map((ev) => ev.user).includes(ev.organizer))) {
      return false;
    }

    for (const f in selectedFilters) {
      for (const cat in ev.categories) {
        if (selectedFilters[f] === ev.categories[cat]) {
          filtered = true;
        }
      }
    }

    if (filtered || selectedFilters.length === 0) {
      let arr = [];
      for (const f in friendFilters) {
        if (friendFilters[f] === "attending") { 
          arr = curUser.attending;
          if (arr.includes(ev._id)) {
            filtered = true;
          } else {
            filtered = false;
          }
        }
        if (friendFilters[f] === "invited") { 
          arr = curUser.invited;
          if (arr.includes(ev._id)) {
            filtered = true;
          } else {
            filtered = false;
          }
        }
      }
    }

    return filtered;
  }

  return (
    <div className='bg-stone-100 min-h-screen'>
      <div className='mx-auto flex flex-col items-center justify-center h-full'>
        <Header icons={true} curUser={curUser} setCurUser={setCurUser}/>
        <div className={'mt-5 w-11/12 flex flex-row flex-nowrap justify-between'}>
          <div className='flex flex-row'>
            <Switch
              checked={listActive}
              onChange={setListActive}
              className={`${
                listActive ? 'bg-blue-600' : 'bg-gray-400'
              } relative inline-flex h-6 w-11 items-center rounded-full`}
            >
              <span
                className={`${
                  listActive ? 'translate-x-6' : 'translate-x-1'
                } inline-block h-4 w-4 transform rounded-full bg-white transition`}
              />
            </Switch>
            <span className='pl-2'>List View</span>
          </div>
          <div className='flex flex-row'>
            <Switch
              checked={friendEventsActive}
              onChange={setFriendEventsActive}
              className={`${
                friendEventsActive ? 'bg-blue-600' : 'bg-gray-400'
              } relative inline-flex h-6 w-11 items-center rounded-full`}
            >
              <span
                className={`${
                  friendEventsActive ? 'translate-x-6' : 'translate-x-1'
                } inline-block h-4 w-4 transform rounded-full bg-white transition`}
              />
            </Switch>
            <span className='pl-2'>Events Hosted By Your Friends Only</span>
          </div>
          <div className='justify-end content-end items-end right-0'>
            <div className='flex flew-row flex-nowrap'>
              <CategoryFilter selectedFilters={selectedFilters} setSelectedFilters={setSelectedFilters} />
              <FriendFilter friendFilters={friendFilters} setFriendFilters={setFriendFilters} />
            </div>
          </div>
        </div>
        {listActive ? (
          <div className='my-5 w-11/12 md:grid md:grid-cols-3 items-center justify-center'>
            {[...eventList, ...privateEventList].filter((ev) => {
              return toDisplayEvent(ev);
            })
            .map((event) => {
              return (
                <EventCard key={event._id} event={event}/>
              );
            })}
          </div>
          ) : (
          <div className='my-3 mb-5 w-11/12 items-center justify-center'>
            <Map events={[...eventList, ...privateEventList].filter((ev) => {
              return toDisplayEvent(ev);
            })} />
        </div>
        )}
      </div>
      
    </div>
  );
  
}

export default EventList;