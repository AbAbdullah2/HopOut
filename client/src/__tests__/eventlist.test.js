import '@testing-library/jest-dom'
import {render, screen} from '@testing-library/react'
import renderer from 'react-test-renderer';
import EventList from '../pages/EventList'
import { getAllEvents } from '../services/api';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

test('matches snapshot', async () => {
  const tree = await renderer.create(<EventList/>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders proper event data', async () => {
  render(<EventList />);

  let eventData = [];

  getAllEvents().then((res) => {
    eventData = res.data.data;
  });

  eventData.forEach((event) => {
    expect(screen.getByText(event.event.title)).toBeInTheDocument();
  });
});