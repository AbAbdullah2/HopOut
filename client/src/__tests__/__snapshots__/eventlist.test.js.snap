// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="bg-stone-100 min-h-screen"
>
  <div
    className="mx-auto flex flex-col items-center justify-center h-full"
  >
    <div
      className="w-full bg-slate-800 p-5 text-white flex flex-row flex-nowrap justify-between"
    >
      <a
        className="mx-2"
        href="/"
      >
        <span
          className="text-4xl font-extrabold italic text-blue-400 mr-1"
        >
          H
        </span>
        <span
          className="text-4xl"
        >
          opout
        </span>
      </a>
      <div
        className="flex justify-end content-end items-end right-0 text-2xl font-semibold space-x-4 mx-2 mr-4"
      >
        <a
          className="hover:text-blue-400"
          href="/events"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-calendar px-1"
            data-icon="calendar"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <span
            className="pl-1 invisible hidden md:visible md:inline"
          >
            Events
          </span>
        </a>
        <a
          className="hover:text-blue-400"
          href="/create"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plus px-1"
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M240 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H32c-17.7 0-32 14.3-32 32s14.3 32 32 32H176V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H384c17.7 0 32-14.3 32-32s-14.3-32-32-32H240V80z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <span
            className="pl-1 invisible hidden md:visible md:inline"
          >
            Create
          </span>
        </a>
        <a
          className="hover:text-blue-400"
          href="/account"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-user px-1"
            data-icon="user"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <span
            className="pl-1 invisible hidden md:visible md:inline"
          >
            Profile
          </span>
        </a>
      </div>
    </div>
    <div
      className="my-5 w-11/12 md:grid md:grid-cols-3 items-center justify-center"
    />
  </div>
</div>
`;
